ot3_state_manager/ot3_state_manager/ot3_state_manager.py

```diff
@@ -0,0 +1,262 @@
+        pos_dict = self._pos_dict()
+
+        # Ignoring the mypy error because we are checking that the value is None
+        return {
+            key: (pos_dict[key].commanded_position if value is not None else None)  # type: ignore[union-attr] # noqa: E501
+            for key, value in pos_dict.items()
+        }
+
+    @property
+    def encoder_position(self) -> Dict[str, Optional[float]]:
+        """Returns dictionary with axis mapped to encoder position of said axis.
+
+        If hardware is not attached for axis, then position value will be None.
+        """
+        pos_dict = self._pos_dict()
+
+        # Ignoring the mypy error because we are checking that the value is None
+        return {
+            key: (pos_dict[key].encoder_position if value is not None else None)  # type: ignore[union-attr] # noqa: E501
+            for key, value in pos_dict.items()
+        }
+
+    @property
+    def updatable_positions(self) -> List[UpdatablePositions]:
+        """Returns a list of positions that are available to be updated."""
+        return [
+            key
+            for key, value in self._updatable_pos_to_hardware_dict.items()
+            if value is not None
+        ]
+
+    @property
+    def _updatable_pos_to_hardware_dict(
+        self,
+    ) -> Dict[UpdatablePositions, Optional[Hardware]]:
+
+        return {
+            UpdatablePositions.GANTRY_X: self.gantry_x,
+            UpdatablePositions.GANTRY_Y: self.gantry_y,
+            UpdatablePositions.GRIPPER_MOUNT: self.head,
+            UpdatablePositions.LEFT_PIPETTE_MOUNT: self.head,
+            UpdatablePositions.RIGHT_PIPETTE_MOUNT: self.head,
+            UpdatablePositions.GRIPPER_EXTEND: (
+                self.gripper if self.is_gripper_attached else None
+            ),
+            UpdatablePositions.LEFT_PIPETTE_PLUNGER: (
+                self.left_pipette if self.is_left_pipette_attached else None
+            ),
+            UpdatablePositions.RIGHT_PIPETTE_PLUNGER: (
+                self.right_pipette if self.is_right_pipette_attached else None
+            ),
+        }
+
+    @property
+    def is_gripper_attached(self) -> bool:
+        """Returns True if gripper is attached, False if not."""
+        return self.gripper is not None
+
+    @property
+    def is_left_pipette_attached(self) -> bool:
+        """Returns True if left pipette is attached, False if not."""
+        return self.left_pipette is not None
+
+    @property
+    def is_right_pipette_attached(self) -> bool:
+        """Returns True if right pipette is attached, False if not."""
+        return self.right_pipette is not None
+
+    @property
+    def is_high_throughput_pipette_attached(self) -> bool:
+        """Returns True if high throughput pipette is attached, False if not."""
+        return (
+            self.right_pipette is not None
+            and self.right_pipette.model == PipetteModel.MULTI_96_1000
+        )
+
+    @staticmethod
+    def _update_pos(
```

Comment:
the position update seems a little complex - `_update_pos` and therefore `update_position` can do a couple different things depending how it's called.

I think it's a better idea to have one way to update position that requires everything. That's easier to typecheck and verify the correct use of - just by removing the `Optional`, we can remove the need for `NoPositionProvidedError`. Also, the decision about what to update is ultimately not going to rest here, so let's keep the logic of what to update and what not to update above.

Reply:
eb45b5ae0f393ab82db1b12df7ba201b46b515a0
How's this?